api/src/main/java/org/pmiops/workbench/db/dao/UserService.java

```diff
@@ -341,92 +367,88 @@ public void setDataUseAgreementNameOutOfDate(String newGivenName, String newFami
+        BypassTimeTargetProperty.COMPLIANCE_TRAINING_BYPASS_TIME);
   }
 
-  public DbUser setEmailVerificationBypassTime(Long userId, Timestamp bypassTime) {
-    DbUser user = userDao.findUserByUserId(userId);
-    return updateUserWithRetries(
-        (u) -> {
-          u.setEmailVerificationBypassTime(bypassTime);
-          return u;
-        },
-        user);
+  public void setBetaAccessBypassTime(Long userId, Timestamp bypassTime) {
+    setBypassTimeWithRetries(
+        userId,
+        bypassTime,
+        DbUser::setBetaAccessBypassTime,
+        BypassTimeTargetProperty.BETA_ACCESS_BYPASS_TIME);
   }
 
-  public DbUser setEraCommonsBypassTime(Long userId, Timestamp bypassTime) {
-    DbUser user = userDao.findUserByUserId(userId);
-    return updateUserWithRetries(
-        (u) -> {
-          u.setEraCommonsBypassTime(bypassTime);
-          return u;
-        },
-        user);
+  public void setEraCommonsBypassTime(Long userId, Timestamp bypassTime) {
+    setBypassTimeWithRetries(
+        userId,
+        bypassTime,
+        DbUser::setEraCommonsBypassTime,
+        BypassTimeTargetProperty.ERA_COMMONS_BYPASS_TIME);
   }
 
-  public DbUser setIdVerificationBypassTime(Long userId, Timestamp bypassTime) {
-    DbUser user = userDao.findUserByUserId(userId);
-    return updateUserWithRetries(
-        (u) -> {
-          u.setIdVerificationBypassTime(bypassTime);
-          return u;
-        },
-        user);
+  public void setTwoFactorAuthBypassTime(Long userId, Timestamp bypassTime) {
+    setBypassTimeWithRetries(
+        userId,
+        bypassTime,
+        DbUser::setTwoFactorAuthBypassTime,
+        BypassTimeTargetProperty.TWO_FACTOR_AUTH_BYPASS_TIME);
   }
 
-  public DbUser setTwoFactorAuthBypassTime(Long userId, Timestamp bypassTime) {
-    DbUser user = userDao.findUserByUserId(userId);
-    return updateUserWithRetries(
+  /**
+   * Functional bypass time column setter, using retry logic.
+   *
+   * @param userId id of user getting bypassed
+   * @param bypassTime type of bypass
+   * @param setter void-returning method to call to set the particular bypass field. Should
+   *     typically be a method reference on DbUser, e.g.
+   * @param targetProperty BypassTimeTargetProperty enum value, for auditing
+   */
+  private void setBypassTimeWithRetries(
+      long userId,
+      Timestamp bypassTime,
+      BiConsumer<DbUser, Timestamp> setter,
+      BypassTimeTargetProperty targetProperty) {
+    setBypassTimeWithRetries(userDao.findUserByUserId(userId), bypassTime, targetProperty, setter);
+  }
+
+  private void setBypassTimeWithRetries(
+      DbUser dbUser,
+      Timestamp bypassTime,
+      BypassTimeTargetProperty targetProperty,
+      BiConsumer<DbUser, Timestamp> setter) {
+    updateUserWithRetries(
         (u) -> {
-          u.setTwoFactorAuthBypassTime(bypassTime);
+          setter.accept(u, bypassTime);
+          userServiceAuditAdapter.fireAdministrativeBypassTime(
```

Comment:
Curious why you decided to tie the audit event to the retry loop instead of firing once when the user initiates the action.

Do we really need to know if an update was tried several times? I thought the idea was that an audit event should indicate a user's attempt to do something. With this retry loop, we technically can't distinguish between multiple attempts at bypassing versus a chain of failing calls.

Reply:
I was thinking a failure in the retryer would skip firing the audit event. Let me verify that's what's happening.

I'm really confused why this particular code path needs its own custom retry logic. I'd have though we'd just want one retry wrapper for all DB actions. I.e. why is the user table special? 

Also, we don't have that many users. I don't understand how this situation was frequent enough to warrant writing all this retry stuff.

Reply:
I see what you mean. Pulling the audit line out of the retrier and doing it last is the most consistent with the intent of the audit system and the least noisy as well. Good catch.